x/upgrade/abci_test.go

```diff
@@ -185,6 +209,162 @@ func (s *TestSuite) TestPlanStringer() {
+}
+
+func (s *TestSuite) TestContains() {
+	skipUpgradeHeights := s.keeper.ConvertIntArrayToInt64(viper.GetIntSlice(s.FlagUnsafeSkipUpgrade))
+	s.T().Log("case where array contains the element")
+	present := s.keeper.Contains(skipUpgradeHeights, -1)
+	s.Require().True(present)
+
+	s.T().Log("case where array doesn't contain the element")
+	present = s.keeper.Contains(skipUpgradeHeights, 4)
+	s.Require().False(present)
+}
+
+func (s *TestSuite) TestSkipUpgradeSkippingBoth() {
+	newCtx := sdk.NewContext(s.cms, abci.Header{Height: s.ctx.BlockHeight() + 1, Time: time.Now()}, false, log.NewNopLogger())
+	req := abci.RequestBeginBlock{Header: newCtx.BlockHeader()}
+	err := s.handler(s.ctx, SoftwareUpgradeProposal{Title: "prop", Plan: Plan{Name: "test", Height: s.ctx.BlockHeight() + 1}})
+	s.Require().Nil(err)
+
+	var i int64
+
+	s.T().Log("Verify if skip upgrade flag clears upgrade plan in both cases")
+	viper.Set(s.FlagUnsafeSkipUpgrade, []int{int(s.ctx.BlockHeight() + 1), int(s.ctx.BlockHeight() + 10)})
+	s.VerifySet()
+
+	s.VerifyConversion(viper.GetIntSlice(s.FlagUnsafeSkipUpgrade))
+
+	for i = 1; i <= s.ctx.BlockHeight(); i++ {
+		newCtx = newCtx.WithBlockHeight(s.ctx.BlockHeight() + i)
+		s.Require().NotPanics(func() {
+			s.module.BeginBlock(newCtx, req)
+		})
+	}
+
+	s.T().Log("Verify a second proposal also is being cleared")
+	err = s.handler(s.ctx, SoftwareUpgradeProposal{Title: "prop2", Plan: Plan{Name: "test2", Height: s.ctx.BlockHeight() + 10}})
+	s.Require().Nil(err)
+	s.Require().NotPanics(func() {
+		s.module.BeginBlock(newCtx, req)
+	})
+
+	//To ensure verification is being done only after both uprades are cleared
+	s.T().Log("Verify if both proposals are cleared")
+	s.VerifyCleared(s.ctx)
+	s.VerifyNotDone(s.ctx, "test")
+	s.VerifyNotDone(s.ctx, "test2")
+}
+
+func (s *TestSuite) TestSkipUpgradeSkippingOne() {
+	newCtx := sdk.NewContext(s.cms, abci.Header{Height: s.ctx.BlockHeight() + 1, Time: time.Now()}, false, log.NewNopLogger())
+	req := abci.RequestBeginBlock{Header: newCtx.BlockHeader()}
+	err := s.handler(s.ctx, SoftwareUpgradeProposal{Title: "prop", Plan: Plan{Name: "test", Height: s.ctx.BlockHeight() + 1}})
+	s.Require().Nil(err)
+
+	var i int64
+
+	s.T().Log("Verify if skip upgrade flag clears upgrade plan in one case and does upgrade on another")
+	viper.Set(s.FlagUnsafeSkipUpgrade, []int{int(s.ctx.BlockHeight() + 1)})
+	s.VerifySet()
+
+	s.VerifyConversion(viper.GetIntSlice(s.FlagUnsafeSkipUpgrade))
+
+	for i = 1; i < s.ctx.BlockHeight(); i++ {
+		//To execute begin block multiple times with a new block height
+		newCtx = newCtx.WithBlockHeight(s.ctx.BlockHeight() + i)
+		s.Require().NotPanics(func() {
+			s.module.BeginBlock(newCtx, req)
+		})
+	}
+
+	s.T().Log("Verify the second proposal is not skipped")
+	err = s.handler(s.ctx, SoftwareUpgradeProposal{Title: "prop2", Plan: Plan{Name: "test2", Height: s.ctx.BlockHeight() + 10}})
+	s.Require().Nil(err)
+	newCtx = newCtx.WithBlockHeight(newCtx.BlockHeight() + 1)
```

Comment:
This is very confusing.`newCtx` keeps getting updated in the for loop, so no idea what the height is, relative to `s.ctx.BlockHeight()`. 

For these ones where you test a lot, you can simply define `startHeight` above, and make everything relative to that. (Or use `s.ctx.BlockHeight()` instead of `startHeight`). 

I mean, I would expect this test to pass when this line read `newCtx = s.ctx.WithBlockHeight(s.ctx.BlockHeight() + 1)`